drivers/sensor/max17055/max17055.c

```diff
@@ -0,0 +1,244 @@
+		break;
+	case SENSOR_CHAN_GAUGE_AVG_CURRENT: {
+		int cap_ma;
+
+		cap_ma = capacity_to_ma(config->rsense_mohms,
+					priv->avg_current);
+		set_millis(val, cap_ma);
+		break;
+	}
+	case SENSOR_CHAN_GAUGE_STATE_OF_CHARGE:
+		val->val1 = priv->state_of_charge / 256;
+		val->val2 = priv->state_of_charge % 256 * 1000000 / 256;
+		break;
+	case SENSOR_CHAN_GAUGE_TEMP:
+		val->val1 = priv->internal_temp / 256;
+		val->val2 = priv->internal_temp % 256 * 1000000 / 256;
+		break;
+	case SENSOR_CHAN_GAUGE_FULL_CHARGE_CAPACITY:
+		tmp = capacity_to_ma(config->rsense_mohms, priv->full_cap);
+		set_millis(val, tmp);
+		break;
+	case SENSOR_CHAN_GAUGE_REMAINING_CHARGE_CAPACITY:
+		tmp = capacity_to_ma(config->rsense_mohms, priv->remaining_cap);
+		set_millis(val, tmp);
+		break;
+	case SENSOR_CHAN_GAUGE_TIME_TO_EMPTY:
+		/* Get time in ms */
+		if (priv->time_to_empty == 0xffff) {
+			val->val1 = 0;
+			val->val2 = 0;
+		} else {
+			tmp = priv->time_to_empty * 5625;
+			set_millis(val, tmp);
+		}
+		break;
+	case SENSOR_CHAN_GAUGE_TIME_TO_FULL:
+		if (priv->time_to_full == 0xffff) {
+			val->val1 = 0;
+			val->val2 = 0;
+		} else {
+			/* Get time in ms */
+			tmp = priv->time_to_full * 5625;
+			set_millis(val, tmp);
+		}
+		break;
+	case SENSOR_CHAN_GAUGE_CYCLE_COUNT:
+		val->val1 = priv->cycle_count / 100;
+		val->val2 = priv->cycle_count % 100 * 10000;
+		break;
+	case SENSOR_CHAN_GAUGE_NOM_AVAIL_CAPACITY:
+		tmp = capacity_to_ma(config->rsense_mohms, priv->design_cap);
+		set_millis(val, tmp);
+		break;
+	case SENSOR_CHAN_GAUGE_DESIGN_VOLTAGE:
+		set_millis(val, config->design_voltage);
+		break;
+	case SENSOR_CHAN_GAUGE_DESIRED_VOLTAGE:
+		set_millis(val, config->desired_voltage);
+		break;
+	case SENSOR_CHAN_GAUGE_DESIRED_CHARGING_CURRENT:
+		val->val1 = config->desired_charging_current;
+		val->val2 = 0;
+		break;
+	default:
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static int max17055_sample_fetch(struct device *dev, enum sensor_channel chan)
+{
+	struct max17055_data *priv = dev->driver_data;
+	int rc = 0;
+
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_ALL);
+
+	rc = max17055_reg_read(priv, VCELL, &priv->voltage);
+	if (rc == 0) {
+		rc = max17055_reg_read(priv, AVG_CURRENT, &priv->avg_current);
+	}
+	if (rc == 0) {
+		rc = max17055_reg_read(priv, REP_SOC, &priv->state_of_charge);
+	}
+	if (rc == 0) {
+		rc = max17055_reg_read(priv, INT_TEMP, &priv->internal_temp);
+	}
+	if (rc == 0) {
+		rc = max17055_reg_read(priv, REP_CAP, &priv->remaining_cap);
+	}
+	if (rc == 0) {
+		rc = max17055_reg_read(priv, FULL_CAP_REP, &priv->full_cap);
+	}
+	if (rc == 0) {
+		rc = max17055_reg_read(priv, TTE, &priv->time_to_empty);
+	}
+	if (rc == 0) {
+		rc = max17055_reg_read(priv, TTF, &priv->time_to_full);
+	}
+	if (rc == 0) {
+		rc = max17055_reg_read(priv, CYCLES, &priv->cycle_count);
+	}
+	if (rc == 0) {
+		rc = max17055_reg_read(priv, DESIGN_CAP, &priv->design_cap);
+	}
+	if (rc != 0) {
+		LOG_ERR("Failed to read channel %d", chan);
+		return rc;
+	}
```

Comment:
maybe a cleaner way to write this is winding it up into a loop:

```
struct {
        int reg_addr;
        int16_t *dest;
} regs[] = {
        { VCELL, &priv->voltage },
        { AVG_CURRENT, &priv->avg_current },
        ...
};

for (size_t i = 0; i < ARRAY_SIZE(regs); i++) {
        if ((rc = max17055_reg_read(priv, regs[i].reg_addr, regs[i].dest)) != 0)
                LOG_ERR(...);
                return rc;
        }
}

return 0;
```

Reply:
But some are of different types (uint16_t and int16_t) so that would need casts. I think people wouldn't like that.

Reply:
but the reg_read function only takes `int16_t`? I don't think the cast would be so criminal, as it was just being hidden before by an implicit conversion on the function call.

Reply:
I had missed that the implicit conversion is still done with your code, so it is safe. Changed, thanks.